### DISCUSSION TITLE
상호작용 버튼 만들고싶습니다

### DISCUSSION THREAD

**Discussion by su0833** [naverz/zepeto-studio-kor#442] [5. Q&A]
안녕하세요 제페토에서 상호작용 가능한 버튼을 만들고싶은데(
제가 이쪽엔 지식이 하나도 없는 초보라 일단 무작정 따라하면서 작업하고 있는데 
문제점이 많아 여쭤봅니다

![1](https://user-images.githubusercontent.com/100546192/155929319-fd2d61fe-b696-4417-857f-1c2b7ff28677.png)

일단 첫번째 문제는 위처럼 버튼이 캐릭터를 항상 따라 다니는데, 
제가 원하는건 특정한 오브젝트에 가까이 가면 버튼이 나타나는 걸 원합니다.

![2](https://user-images.githubusercontent.com/100546192/155929301-a6a1ccaf-f9ab-4571-847e-99444ca77706.png)

두번째는 버튼을 클릭하면 캐릭터가 마지막으로 서 있던 자리에서 애니가 적용됩니다ㅠ
저는 버튼을 누르면 캐릭터의 위치에 상관없이 저 의자에 딱 고정되어 앉을 수 있게(정해진 위치값에 앉을 수 있게) 만들고 싶습니다.

세번째는 애니메이션이 끝나면 멈추지 말고 계속 유지하면서 캐릭터 이동키를 눌러서 이동해야 애니메이션이 멈출 수 있게 제작하고 싶습니다.
 
![3](https://user-images.githubusercontent.com/100546192/155930134-b146add6-3c0f-4eee-96ea-0a2e379dc9b0.png)

어떻게해야 원하는 버튼을 만들 수 있을까요ㅠ 
자세히 알려주시면 감사하겠습니다

**Comment by Zepeto-TechEvangelist** [comment]
안녕하세요, 특정 위치에만 버튼을 나타나게 하고 싶다면 버튼의 부모 Canvas를 World Space 속성으로 바꾸시면 됩니다. 
단, 이때 Event Camera를 제페토 카메라로 등록해야 클릭이 가능한데요.
에디터 상에서 제페토 카메라를 끌어다 놓는 것은 현재로서는 불가능하며 코드로 세팅해야 합니다.
제페토 카메라 세팅 코드는 다음과 같습니다. 
아래 내용 참고하셔서 우선 버튼 기능을 완성하신 후 차근차근 기능을 시험해 보세요~

<img width="314" alt="스크린샷 2022-02-28 오후 2 59 39" src="https://user-images.githubusercontent.com/100118650/155932131-8d686bd8-4e25-4e24-a856-5ba1e071844d.png">
<img width="824" alt="스크린샷 2022-02-28 오후 3 09 45" src="https://user-images.githubusercontent.com/100118650/155933091-08dbe7e6-c9a6-4714-baca-303f3305a9bd.png">

**Reply by su0833** [reply]
와우 해결했습니다!! 감사합니다^^

**Reply by shamKE** [reply]
와아.. 덕분에 저도 도움을 받았습니다. 감사합니다!

**Reply by chlguswls32** [reply]
감사합니다!! ㅎㅎ

**Comment by shlifedev** [comment]
### - 상호작용 UI 구현

안녕하세요. 아래 [구글드라이브 링크](https://drive.google.com/file/d/1QXdFIYVeWMGhZysEzVBWZ2wJibQ7ZVbT/view?usp=sharing)는 직접 커스텀으로 상호작용을 만드는 방법에 대한 간단한 샘플 프로젝트 파일입니다. 

위 @Zepeto-TechEvangelist  님께서 설명해주신 캔버스 세팅 내용과 더불어 스크립트 예제까지 포함되어 있습니다.
`InteractableRewardBox.ts` `InteractionObject.ts` `Billboard.ts` 스크립트를 참고하시면 되겠습니다 😀 

drive link : https://drive.google.com/file/d/1QXdFIYVeWMGhZysEzVBWZ2wJibQ7ZVbT/view?usp=sharing



![gif](https://github.com/shlifedev/RobloxOOP/blob/main/Animation.gif?raw=true)
 
 
### - 애니메이션 적용 위치 관련답변

 단순히 애니메이션만 재생 하는 경우 캐릭터는 현재 위치에서 애니메이션이 재생 되는것이 맞습니다.
 
 하지만 만약 의자에 앉는 애니메이션 재생과 동시에 캐릭터의 위치를 의자에 맞춰 주고 싶은경우라면  
 
 현재는 직접 제페토 스크립트를 작성하여 애니메이션 재생 시 캐릭터의 위치 `(transform.position)` 와 회전값`(transform.rotation)` 을 앉아야하는 의자 위치에 맞게 스크립트로 직접 제어하여 이동하여 주셔야 합니다.

**Reply by Park-sechang** [reply]
감사합니다!
혹시 InteractionObject 스크립트는 어떤 오브젝트에 할당되어 있는지 알 수 있을까요?

**Reply by shlifedev** [reply]
@Challenge000 안녕하세요, `InteractionObject.ts`  자체는 오브젝트에 따로 부착되어 있지 않습니다!

`RewardBox` 오브젝트 하위의 `IntertableRewardBox(Detector)` 오브젝트에 컴포넌트로 부착되어 있습니다. 
( `InteractableRewardBox.ts `는 `InteractionObject.ts`를 상속받아  동작합니다!)

하이라키에 `IntertableRewardBox(Detector)` 를 검색하시면 찾아보실 수 있을거에요 ! 😀
![](https://i.imgur.com/6RyVXF8.png)

![](https://i.imgur.com/uT3IsP7.png)

InteractableRewardBox.ts 
```ts
export default class InteractionReawrdBox extends InteractionObject {
    public animator : Animator;

    //상호작용 버튼
    DoInteraction() : void{
         if(this._isTocuhed){
            this.animator.SetTrigger("Toggle");
         }
    } 
}

**Reply by su0833** [reply]
해결 했습니다ㅠ 정말 감사합니다!!

**Reply by hipsc159** [reply]
multiplay에서도 작동하게 하려면 어떻게 해야하나요?
